test/framework/ownerreference_helpers.go

```diff
@@ -0,0 +1,335 @@
+	},
+	dockerMachineTemplateKind: func(owners []metav1.OwnerReference) error {
+		// Base DockerMachineTemplates referenced in a ClusterClass must be owned by the ClusterClass.
+		// DockerMachineTemplates created for specific Clusters in the Topology controller must be owned by a Cluster.
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK}, []schema.GroupVersionKind{clusterClassGVK})
+	},
+	dockerClusterKind: func(owners []metav1.OwnerReference) error {
+		// DockerCluster must be owned by a Cluster.
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
+	},
+	dockerClusterTemplateKind: func(owners []metav1.OwnerReference) error {
+		// DockerClusterTemplate must be owned by a ClusterClass.
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterClassGVK})
+	},
+}
+
+func hasExactOwnersByGVK(refList []metav1.OwnerReference, wantGVKs []schema.GroupVersionKind) error {
+	refGVKs := []schema.GroupVersionKind{}
+	for _, ref := range refList {
+		refGVK, err := ownerRefGVK(ref)
+		if err != nil {
+			return err
+		}
+		refGVKs = append(refGVKs, refGVK)
+	}
+	sort.SliceStable(refGVKs, func(i int, j int) bool {
+		return refGVKs[i].String() > refGVKs[j].String()
+	})
+	sort.SliceStable(wantGVKs, func(i int, j int) bool {
+		return wantGVKs[i].String() > wantGVKs[j].String()
+	})
+	if !reflect.DeepEqual(wantGVKs, refGVKs) {
+		return fmt.Errorf("wanted %v, actual %v", wantGVKs, refGVKs)
+	}
+	return nil
+}
+
+// NOTE: we are using hasOneOfExactOwnersByGVK as a convenience approach for checking owner references on objects that
+// can have different owner references depending on the cluster topology.
+// In a follow-up iteration we can make improvements to check owner references according to the specific use cases vs checking generically "oneOf".
+func hasOneOfExactOwnersByGVK(refList []metav1.OwnerReference, possibleGVKS ...[]schema.GroupVersionKind) error {
+	var allErrs []error
+	for _, wantGVK := range possibleGVKS {
+		err := hasExactOwnersByGVK(refList, wantGVK)
+		if err != nil {
+			allErrs = append(allErrs, err)
+			continue
+		}
+		return nil
+	}
+	return kerrors.NewAggregate(allErrs)
+}
+
+func ownerRefGVK(ref metav1.OwnerReference) (schema.GroupVersionKind, error) {
+	refGV, err := schema.ParseGroupVersion(ref.APIVersion)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+	return schema.GroupVersionKind{Version: refGV.Version, Group: refGV.Group, Kind: ref.Kind}, nil
+}
+
+func setClusterPause(ctx context.Context, cli client.Client, clusterKey types.NamespacedName, value bool) {
+	cluster := &clusterv1.Cluster{}
+	Expect(cli.Get(ctx, clusterKey, cluster)).To(Succeed())
+
+	unpausePatch := client.RawPatch(types.MergePatchType, []byte(fmt.Sprintf("{\"spec\":{\"paused\":%v}}", value)))
+	Expect(cli.Patch(ctx, cluster, unpausePatch)).To(Succeed())
+}
+
+func forceClusterClassReconcile(ctx context.Context, cli client.Client, clusterKey types.NamespacedName) {
```

Comment:
Q: Wouldn't it make also sense to do something for the non-ClusterClass case and thus just have one implementation modifying the cluster?

Or do we only have a delay in the ClusterClass case because of the way watches are set up?

Depending on answers, would be good to have a short comment

Reply:
Yeah - this is down to the way ClusterClass ownerReferences for templates are re-reconciled. CC controller doesn't watch the templates, so it's down to the sync period here.